File: python/examples/core/03-widget-crud/main.py
"""Example 03: full CRUD lifecycle for a note widget.

Runs six ordered steps against `CANVUS_CANVAS_ID`. On any step's failure the
program logs the error and STOPS — subsequent steps would only confuse the
debugging story. The lifecycle:

    1. Create a note at (100, 100).
    2. Print the new widget id.
    3. PATCH the note's text and background colour.
    4. Print the updated widget.
    5. DELETE the note.
    6. Verify the delete by GET — expecting :class:`NotFoundError`.
"""

from __future__ import annotations

import asyncio
import sys
from datetime import UTC, datetime

import structlog
from pydantic import Field, ValidationError
from pydantic_settings import BaseSettings, SettingsConfigDict

from canvus_sdk import APIError, Client, NotFoundError, configure_logging

logger = structlog.get_logger(__name__)


class WidgetCRUDSettings(BaseSettings):
    """Environment configuration for the widget CRUD example."""

    model_config = SettingsConfigDict(
        env_prefix="CANVUS_",
        env_file=".env",
        env_file_encoding="utf-8",
        extra="ignore",
    )

    api_url: str = Field(...)
    api_key: str = Field(...)
    canvas_id: str = Field(..., description="Canvas to operate on.")


async def main() -> int:
    """Entry point."""
    configure_logging()
    try:
        settings = WidgetCRUDSettings()  # type: ignore[call-arg]
    except ValidationError as e:
        logger.error("config invalid", errors=e.errors())
        return 2

    async with Client(settings.api_url, settings.api_key) as client:
        canvas_id = settings.canvas_id

        # Step 1: create
        now_iso = datetime.now(UTC).isoformat()
        try:
            created = await client.widgets.notes.create(
                canvas_id,
                {
                    "text": f"hello from python @ {now_iso}",
                    "location": {"x": 100.0, "y": 100.0},
                    "size": {"width": 250.0, "height": 200.0},
                },
            )
        except APIError as e:
            logger.error(
                "step1_create failed",
                status_code=e.status_code,
                response_body=e.response_body,
            )
            return 1
        note_id = created.id
        if note_id is None:
            logger.error("step1_create returned no id", note=created.model_dump())
            return 1
        logger.info("step1_create ok", note_id=note_id, text=created.text)

        # Step 2: print
        print(f"created note id: {note_id}")

        # Step 3: patch
        try:
            updated = await client.widgets.notes.update(
                canvas_id,
                note_id,
                {
                    "text": f"updated from python @ {datetime.now(UTC).isoformat()}",
                    "background_color": "#3aaa34ff",
                },
            )
        except APIError as e:
            logger.error(
                "step3_patch failed",
                status_code=e.status_code,
                response_body=e.response_body,
            )
            return 1
        logger.info(
            "step3_patch ok",
            note_id=updated.id,
            text=updated.text,
            background_color=updated.background_color,
        )

        # Step 4: print updated
        print(f"updated note: text={updated.text!r} bg={updated.background_color}")

        # Step 5: delete
        try:
            await client.widgets.notes.delete(canvas_id, note_id)
        except APIError as e:
            logger.error(
                "step5_delete failed",
                status_code=e.status_code,
                response_body=e.response_body,
            )
            return 1
        logger.info("step5_delete ok", note_id=note_id)

        # Step 6: verify by GET → expect NotFoundError
        try:
            ghost = await client.widgets.notes.get(canvas_id, note_id)
            logger.error(
                "step6_verify FAILED — note still exists after delete",
                note=ghost.model_dump(),
            )
            return 1
        except NotFoundError:
            logger.info("step6_verify ok", note_id=note_id, status_code=404)
            print(f"verified deletion of note {note_id} (NotFoundError as expected)")
        except APIError as e:
            logger.error(
                "step6_verify got non-404 error",
                status_code=e.status_code,
                response_body=e.response_body,
            )
            return 1

    return 0


if __name__ == "__main__":
    sys.exit(asyncio.run(main()))

File: go/examples/core/03-widget-crud/main.go
// Command 03-widget-crud exercises the full lifecycle of a single sticky
// note on a Canvus canvas: create, update, delete, then verify the delete
// took effect.
//
// It exists to demonstrate the idiomatic shape of widget operations through
// the SDK and to give SDK consumers a copy-pasteable template for their own
// widget code.
package main

import (
	"context"
	"errors"
	"fmt"
	"log/slog"
	"os"
	"time"

	"github.com/jaypaulb/MT-Canvus-Tools/go/sdk/canvus"
)

func main() {
	setupLogging()
	if err := run(); err != nil {
		slog.Error("widget-crud failed", "err", err)
		os.Exit(1)
	}
}

func run() error {
	baseURL, err := mustEnv("CANVUS_BASE_URL")
	if err != nil {
		return err
	}
	apiKey, err := mustEnv("CANVUS_API_KEY")
	if err != nil {
		return err
	}
	canvasID, err := mustEnv("CANVUS_CANVAS_ID")
	if err != nil {
		return err
	}

	cfg := &canvus.SessionConfig{BaseURL: baseURL}
	s := canvus.NewSession(cfg, canvus.WithAPIKey(apiKey))
	ctx := context.Background()

	// Step 1 — create.
	createText := fmt.Sprintf("hello from Go @ %s", time.Now().UTC().Format(time.RFC3339))
	createReq := map[string]any{
		"widget_type": "note",
		"text":        createText,
	}
	note, err := s.CreateNote(ctx, canvasID, createReq)
	if err != nil {
		return fmt.Errorf("step 1 CreateNote: %w", err)
	}
	slog.Info("step 1 — note created",
		"widget_id", note.ID,
		"text", note.Text,
		"background_color", note.BackgroundColor)

	// Step 2 — print id (already logged; emit to stdout for piping).
	fmt.Printf("created note id: %s\n", note.ID)

	// Step 3 — update.
	updateText := fmt.Sprintf("updated from Go @ %s", time.Now().UTC().Format(time.RFC3339))
	updateReq := map[string]any{
		"widget_type":      "note",
		"text":             updateText,
		"background_color": "#3AAA34FF",
	}
	updated, err := s.UpdateNote(ctx, canvasID, note.ID, updateReq)
	if err != nil {
		return fmt.Errorf("step 3 UpdateNote: %w", err)
	}
	slog.Info("step 3 — note updated",
		"widget_id", updated.ID,
		"text", updated.Text,
		"background_color", updated.BackgroundColor)

	// Step 4 — print updated widget (already logged; emit to stdout).
	fmt.Printf("updated note id: %s text: %q\n", updated.ID, updated.Text)

	// Step 5 — delete.
	if err := s.DeleteNote(ctx, canvasID, note.ID); err != nil {
		return fmt.Errorf("step 5 DeleteNote: %w", err)
	}
	slog.Info("step 5 — note deleted", "widget_id", note.ID)

	// Step 6 — verify deletion. The server should return 404; the SDK wraps
	// that as canvus.ErrNotFound via APIError.Unwrap.
	_, err = s.GetNote(ctx, canvasID, note.ID)
	switch {
	case err == nil:
		return fmt.Errorf("step 6 expected NotFound after delete, got nil error")
	case errors.Is(err, canvus.ErrNotFound):
		slog.Info("step 6 — verified deletion (got ErrNotFound as expected)")
		return nil
	default:
		return fmt.Errorf("step 6 GetNote returned unexpected error: %w", err)
	}
}

// mustEnv returns the value of name or an error if it is empty/unset.
func mustEnv(name string) (string, error) {
	v := os.Getenv(name)
	if v == "" {
		return "", fmt.Errorf("missing required env var %s", name)
	}
	return v, nil
}

// setupLogging configures slog.Default with a text or JSON handler.
func setupLogging() {
	var h slog.Handler
	switch os.Getenv("LOG_FORMAT") {
	case "json":
		h = slog.NewJSONHandler(os.Stderr, &slog.HandlerOptions{Level: slog.LevelInfo})
	default:
		h = slog.NewTextHandler(os.Stderr, &slog.HandlerOptions{Level: slog.LevelInfo})
	}
	slog.SetDefault(slog.New(h))
}
